File: 프로그래머스/1/172928. 공원 산책/공원 산책.py
def solution(park, routes):
    # 시작 위치 찾기
    start = []
    for row_n, row in enumerate(park):
        if 'S' in row:
            col_n = row.index('S')
            start = [row_n, col_n]
    print('시작 위치 : ', start)
    for route in routes:
        op, n = route.split()
        temp_location = start[:]        # 이동했을 때의 포인터
        
        for _ in range(int(n)):
            temp_location = move(temp_location, op)
            if (temp_location[0] < 0 or temp_location[0] >= len(park)
                    or temp_location[1] < 0 or temp_location[1] >= len(park[0])
                    or park[temp_location[0]][temp_location[1]] == 'X'
                   ):
                    temp_location = start[:]
                    break
        start = temp_location[:]
            
    return start

def move(temp_location, op):
    if op == 'E':
        temp_location[1] += 1
    elif op == 'W':
        temp_location[1] -= 1
    elif op == 'S':
        temp_location[0] += 1
    else:
        temp_location[0] -= 1
    return temp_location

File: 프로그래머스/1/172928. 공원 산책/공원 산책.swift
import Foundation

func solution(_ park:[String], _ routes:[String]) -> [Int] {
    let park = park.map { Array($0) }
    var location = findStartIndex(park)
    let move = ["N": (-1, 0) , "W": (0, -1), "S": (1, 0), "E": (0, 1)]
    
    for route in routes {
        let (direction, count) = parse(route)
        var isSuccess = true
        var (nextRow, nextCol) = (location.0, location.1)

        for _ in 0..<count {
            nextRow += move[direction]!.0
            nextCol += move[direction]!.1

            guard nextRow >= 0 && nextRow < park.count && nextCol >= 0 && nextCol < park[0].count
                    && park[nextRow][nextCol] != "X" else {
                isSuccess = false
                break
            }
        }

        if isSuccess {
            location = (nextRow, nextCol)
        }
    }

    return [location.0, location.1]
}

func findStartIndex(_ park: [[Character]]) -> (Int, Int) {
    for i in 0..<park.count {
        for j in 0..<park[0].count {
            if park[i][j] == "S" {
                return (i, j)
            }
        }
    }
    return (0, 0)
}

func parse(_ command: String) -> (String, Int) {
    let components = command.split(separator: " ")
    return (String(components[0]), Int(components[1])!)
}
